components/core/src/reducer/reducer_server.cpp

```diff
@@ -0,0 +1,670 @@
+            SPDLOG_ERROR(
+                    "Rejecting connection from worker with job_id={} during processing of "
+                    "job_id={}",
+                    job_id,
+                    rctx->ctx->get_job_id()
+            );
+            return;
+        }
+        rctx->bytes_occupied = 0;
+        char const response = 'y';
+        boost::system::error_code e;
+        int transferred = boost::asio::write(rctx->socket, boost::asio::buffer(&response, 1), e);
+        if (e || transferred < sizeof(response)) {
+            SPDLOG_ERROR("Rejecting connection because of connection error while attempting to "
+                         "send acceptance");
+            return;
+        }
+
+        queue_receive_task(std::move(rctx));
+    } else {
+        rctx->bytes_occupied = bytes_remaining;
+        queue_validate_sender_task(std::move(rctx));
+    }
+}
+
+void queue_validate_sender_task(std::shared_ptr<RecordReceiverContext> rctx) {
+    boost::asio::async_read(
+            rctx->socket,
+            boost::asio::buffer(
+                    &rctx->buf[rctx->bytes_occupied],
+                    sizeof(int32_t) - rctx->bytes_occupied
+            ),
+            boost::bind(
+                    validate_sender_task,
+                    boost::asio::placeholders::error,
+                    boost::asio::placeholders::bytes_transferred,
+                    rctx
+            )
+    );
+}
+
+void accept_task(
+        boost::system::error_code const& error,
+        std::shared_ptr<ServerContext> ctx,
+        std::shared_ptr<RecordReceiverContext> rctx
+) {
+    if (false == error.failed() && ServerStatus::RUNNING == ctx->get_status()) {
+        queue_validate_sender_task(std::move(rctx));
+        queue_accept_task(std::move(ctx));
+    } else if (error.failed() && boost::system::errc::operation_canceled == error.value()) {
+        SPDLOG_INFO("Accept task cancelled");
+    } else if (error.failed()) {
+        // tcp acceptor socket was closed -- don't re-queue accept task
+        SPDLOG_ERROR("TCP acceptor socket closed");
+        if (ctx->get_tcp_acceptor().is_open()) {
+            ctx->get_tcp_acceptor().close();
+        }
+    } else {
+        SPDLOG_WARN(
+                "Rejecting connection while not in RUNNING state, state={}",
+                server_status_to_string(ctx->get_status())
+        );
+        queue_accept_task(std::move(ctx));
+    }
+}
+
+void queue_accept_task(std::shared_ptr<ServerContext> ctx) {
+    auto rctx = RecordReceiverContext::NewReceiver(ctx);
+
+    ctx->get_tcp_acceptor().async_accept(
+            rctx->socket,
+            boost::bind(accept_task, boost::asio::placeholders::error, ctx, rctx)
+    );
+}
+
+void poll_db(
+        boost::system::error_code const& e,
+        std::shared_ptr<ServerContext> ctx,
+        boost::asio::steady_timer* poll_timer
+) {
+    if (ServerStatus::IDLE == ctx->get_status()) {
+        ctx->set_status(ctx->execute_assign_new_job());
```

Comment:
Why not set the status within the method?

Reply:
This is half a holdover from when everything was written in a more function-oriented style, and half an attempt at readability. The nice part of doing it this way is that all of the state transitions happen with an explicit set_status inside of this poll_db function instead of implicitly inside of the various members of ServerStatus.

Granted if you're not already aware of what the states are/what kinds of transitions there are this probably isn't much more readable.

Reply:
Mm, keeping the transitions in one method is fair but since the return values are directly passed to set_status, it does hide what the transition actually is. I need to think more if there's a better solution.